Added Seeded Random Number Generation

morsecodist opened the pull request:
I know you are not 100% committed to how you are doing distributions but this works with your current setup. I wanted seeded random number generation for testing. All of your usage is preserved. The additional functionality works as follows:

```
math.seedrandom('abc');
math.random() // Always predicts same thing
```

This works with all of your current random functions so it makes the entire system deterministic with a single call. Calling:

```
math.seedrandom();
```

Seeds the random number generator with a random seed. 

I also updated the unit tests for distribution to use this seed paradigm rather than the way you did it by replacing the random number generator manually at test time.

josdejong commented:
Thanks Todd, a random seed is indeed a must, thanks for your PR. The distribution related code can use some refactoring but it's working fine. 

I will add a few comments to your code, would like to merge your work soon!

josdejong commented on lib/function/probability/rng.js:
```diff
@@ -0,0 +1,16 @@
+      generator = seedrandom(seed)
+    }
+  },
+  generate: () => generator()
```
Sorry, all math.js code is still ES5, no ES6 arrow functions supported.

josdejong commented on lib/function/probability/seedrandom.js:
```diff
@@ -0,0 +1,33 @@
+   *
+   * @param {string} [seed] seed for random number generator
+   */
+  // TODO: rework random to a typed-function
```
Indeed, can you please rewrite to a typed function? Will be something like
```js
var seedrandom = typed('seedrandom', {
  'string': rng.setSeed
});
```

josdejong commented on package.json:
```diff
@@ -76,9 +76,10 @@
     "decimal.js": "5.0.8",
     "fraction.js": "3.3.1",
-    "complex.js": "2.0.1",
+    "seed-random": "^2.2.0",
```
Can you please use an exact version number `"2.2.0"`?

josdejong commented:
Just thinking aloud, instead of a new function `math.seedrandom` we could also create this as a new config option, like

```js
math.config({randomSeed: 'ABC'});   // equivalent to `math.seedrandom('ABC')`
math.config({randomSeed: null});    // equivalent to `math.seedrandom()`
```

When setting it as a config option it may be more clear that you're changing something globally, which may can have effect behavior of various functions. Would that make sense?

morsecodist commented:
The config option makes sense. I just did it this way because it's similar to how numpy's [random.seed](https://docs.scipy.org/doc/numpy/reference/generated/numpy.random.seed.html) function works.

morsecodist commented:
I addressed the comments you brought up. The merge conflict seems to be because I accidentally reordered the dependencies.

josdejong commented:
Cool, thanks. Let's go for the config option then! Can you rewrite your code into a config option? Rougly:

- Add a new value `randomSeed` in the default config object [in core.js](https://github.com/josdejong/mathjs/blob/master/lib/core/core.js#L57-L77).
- In distribution.js create a listener for changes in the config, and on change create a new random generator, something like:
  ```js
  math.on('config', function (curr, prev) {
    if (curr.randomSeed !== prev.randomSeed) {
      //... create new random generator
    }
  });

  ```
- Update [docs](https://github.com/josdejong/mathjs/blob/master/docs/core/configuration.md) and code comments of [function config](https://github.com/josdejong/mathjs/blob/master/lib/core/function/config.js)

The merge conflict should be easy to solve, it's probably also because I updated version numbers of dependencies two days ag.

morsecodist commented:
Two issues with this. 

1. It should allow for users to refresh the seed with the same seed.
2. Doing it asynchronously with a listener creates issues when calling random too close together.

I have a working solution adding both a seed and an `seed-random` instance to the config object then I added a piece to the config function that updates the `seed-random` instance with the seed if one is supplied. Is that alright?

josdejong commented:
1. You're right, well, then there is no need for the `if` statement :)
2. That's a good point. The `math.on('config', ...)` event is actually fired synchronously so it will work, but such a callback structure is normally async (at least as a consumer you must assume it's async). Hm, that's something we have to give more thought. The idea behind this event emitter is that the config function itself is lean and mean and decoupled from code actually using the config, and all parts of math.js that need to be aware of changes in the config can register an event listener.

I forgot one thing: this `math.on` event is available as a fifth argument of the factory function only when you explicitely export the factory function along with a property `exports.math = true`, like in [`import.js`](https://github.com/josdejong/mathjs/blob/master/lib/core/function/import.js#L263).

morsecodist commented:
It currently works how you described but I ran into a few issues so here is what I did.

1. Removing the if statement that checks if the seed was changed makes it so the rng is reset every time config is called even if the user doesn't specify a `randomSeed`. To solve this I modified the config event emission to include the original options supplied by the user. This way I can check if the user specified a `randomSeed` before seeding a new pseudo random number generator. The code still passes all tests with this change and I updated the associated comments as far as I could find them

2. Because of the way distributions return objects I couldn't reset the RNG without a wrapper function. I separated this out into another file using the factory pattern you set up. The wrapper function is necessary and I think keeping it separate makes things a bit cleaner. The new file is `seededRNG.js`. It allowed me to change `distribution.js` as little as possible and I think it makes it easier to update however seeded random number generation is handled in the future. 

I also updated the docs and unit tests related to `config` to work with the new parameter. All tests are passing.

josdejong commented:
Awesome, thanks for the updates!

1. ah, that's a nifty solution by passing the actual changes in the config as third argument! I hadn't realized at all that this issue would arise. Maybe naming the third argument of the callback `changes` instead of `options` would be more clear?
2. That's indeed a good solution. I see currently `seededRNG` is exposed on the `math` namespace, but maybe we want to keep this as an internal function? That can be achieved by removing the line `exports.name = 'seededRNG'`.
  
3. Can you resolve the merge conflict? (I can do it too if you're not sure)

morsecodist commented:
Made the changes you requested. I also added a `toString` call so you can use seeds of any type. In the documentation I still have it as strings since they are the most straightforward type to use. 

I resolved the merge conflicts by using the more recent dependency in each case with conflicting version numbers.

josdejong commented:
Sweat @morsecodist , thanks for your work. I really like the result :)
